File: scripts/apply-slug-mapping.py
#!/usr/bin/env python3
import csv
import os
import re
import json
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
CSV_PATH = ROOT / "data" / "slug-mapping.csv"
OUT_DIR = ROOT / "content" / "blog"
DATA_DIR = ROOT / "data"
BLOGGER_DOMAIN = "https://nepaliessaybook.blogspot.com"


def find_file_by_old_url(out_dir: Path, target_old_url: str) -> str | None:
    if not out_dir.exists():
        return None
    for file in out_dir.iterdir():
        if not file.is_file() or file.suffix != ".mdx":
            continue
        content = file.read_text(encoding="utf-8")
        match = re.search(r'^oldUrl:\s*"?([^"\n]+)"?', content, re.MULTILINE)
        if match and match.group(1).strip() == target_old_url:
            return file.name
    return None


def update_frontmatter_slug(content: str, new_slug: str) -> str:
    return re.sub(
        r'^slug:\s*"?[^"\n]+"?',
        f'slug: "{new_slug}"',
        content,
        count=1,
        flags=re.MULTILINE,
    )


def main():
    rows = []
    with open(CSV_PATH, "r", encoding="utf-8", newline="") as f:
        reader = csv.DictReader(f)
        for row in reader:
            rows.append(row)

    print(f"Loaded {len(rows)} slug mapping rows")

    redirects: dict[str, str] = {}
    renamed = 0
    skipped = 0
    not_found: list[str] = []

    for row in rows:
        old_filename = row["old_filename"]
        old_slug = row["old_slug"]
        new_slug = row["new_slug"]
        old_url = f"{BLOGGER_DOMAIN}{old_filename}"

        current_file = find_file_by_old_url(OUT_DIR, old_url)
        if not current_file:
            not_found.append(old_url)
            continue

        old_path = OUT_DIR / current_file
        content = old_path.read_text(encoding="utf-8")
        new_content = update_frontmatter_slug(content, new_slug)
        new_path = OUT_DIR / f"{new_slug}.mdx"

        if old_path != new_path:
            old_path.write_text(new_content, encoding="utf-8")
            old_path.rename(new_path)
            renamed += 1
            print(f"✓ {old_slug} → {new_slug}")
        else:
            new_path.write_text(new_content, encoding="utf-8")
            skipped += 1
            print(f"  {new_slug} (unchanged)")

        redirects[old_url] = f"/blog/{new_slug}"

        if old_slug != new_slug:
            old_portfolio_path = f"/blog/{old_slug}"
            if (
                old_portfolio_path in redirects
                and redirects[old_portfolio_path] != f"/blog/{new_slug}"
            ):
                print(
                    f"Conflict: {old_portfolio_path} already redirects to "
                    f"{redirects[old_portfolio_path]}, not /blog/{new_slug}"
                )
            else:
                redirects[old_portfolio_path] = f"/blog/{new_slug}"

    (DATA_DIR / "redirects.json").write_text(
        json.dumps(redirects, indent=2), encoding="utf-8"
    )

    print("\nSlug mapping applied:")
    print(f"  Renamed: {renamed}")
    print(f"  Unchanged: {skipped}")
    print(f"  Not found: {len(not_found)}")
    print(f"  Redirect entries: {len(redirects)}")

    if not_found:
        print("\nCould not find files for:")
        for url in not_found[:10]:
            print(f"  - {url}")
        raise SystemExit(1)


if __name__ == "__main__":
    main()

File: scripts/apply-slug-mapping.ts
import * as fs from "fs";
import * as path from "path";

const CSV_PATH = path.resolve(process.cwd(), "data/slug-mapping.csv");
const OUT_DIR = path.resolve(process.cwd(), "content/blog");
const DATA_DIR = path.resolve(process.cwd(), "data");
const BLOGGER_DOMAIN = "https://nepaliessaybook.blogspot.com";

interface MappingRow {
  title: string;
  category: string;
  oldFilename: string;
  oldSlug: string;
  newSlug: string;
  isBadOld: boolean;
  published: string;
}

function parseCsv(content: string): MappingRow[] {
  const lines = content.trim().split("\n");
  const headers = lines[0].split(",").map((h) => h.trim());
  const rows: MappingRow[] = [];

  for (let i = 1; i < lines.length; i++) {
    const line = lines[i];
    const cells: string[] = [];
    let current = "";
    let inQuotes = false;

    for (const char of line) {
      if (char === '"') {
        inQuotes = !inQuotes;
      } else if (char === "," && !inQuotes) {
        cells.push(current.trim());
        current = "";
      } else {
        current += char;
      }
    }
    cells.push(current.trim());

    if (cells.length < 7) continue;

    rows.push({
      title: cells[0],
      category: cells[1],
      oldFilename: cells[2],
      oldSlug: cells[3],
      newSlug: cells[4],
      isBadOld: cells[5].toLowerCase() === "true",
      published: cells[6],
    });
  }

  return rows;
}

function findFileByOldUrl(outDir: string, targetOldUrl: string): string | null {
  if (!fs.existsSync(outDir)) return null;
  for (const file of fs.readdirSync(outDir)) {
    if (!file.endsWith(".mdx")) continue;
    const content = fs.readFileSync(path.join(outDir, file), "utf-8");
    const match = content.match(/^oldUrl:\s*"?([^"\n]+)"?/m);
    if (match?.[1].trim() === targetOldUrl) {
      return file;
    }
  }
  return null;
}

function updateFrontmatterSlug(content: string, newSlug: string): string {
  return content.replace(/^slug:\s*"?[^"\n]+"?/m, `slug: "${newSlug}"`);
}

async function main() {
  const csvContent = fs.readFileSync(CSV_PATH, "utf-8");
  const rows = parseCsv(csvContent);

  console.log(`Loaded ${rows.length} slug mapping rows`);

  const redirects: Record<string, string> = {};
  let renamed = 0;
  let skipped = 0;
  let notFound: string[] = [];

  for (const row of rows) {
    const oldUrl = `${BLOGGER_DOMAIN}${row.oldFilename}`;
    const currentFile = findFileByOldUrl(OUT_DIR, oldUrl);

    if (!currentFile) {
      notFound.push(oldUrl);
      continue;
    }

    const oldFilePath = path.join(OUT_DIR, currentFile);
    const content = fs.readFileSync(oldFilePath, "utf-8");
    const newContent = updateFrontmatterSlug(content, row.newSlug);
    const newFilePath = path.join(OUT_DIR, `${row.newSlug}.mdx`);

    if (oldFilePath !== newFilePath) {
      fs.writeFileSync(oldFilePath, newContent, "utf-8");
      fs.renameSync(oldFilePath, newFilePath);
      renamed++;
      console.log(`✓ ${row.oldSlug} → ${row.newSlug}`);
    } else {
      fs.writeFileSync(newFilePath, newContent, "utf-8");
      skipped++;
      console.log(`  ${row.newSlug} (unchanged)`);
    }

    redirects[oldUrl] = `/blog/${row.newSlug}`;

    if (row.oldSlug !== row.newSlug) {
      const oldPortfolioPath = `/blog/${row.oldSlug}`;
      if (redirects[oldPortfolioPath] && redirects[oldPortfolioPath] !== `/blog/${row.newSlug}`) {
        console.warn(`Conflict: ${oldPortfolioPath} already redirects to ${redirects[oldPortfolioPath]}, not ${row.newSlug}`);
      } else {
        redirects[oldPortfolioPath] = `/blog/${row.newSlug}`;
      }
    }
  }

  fs.writeFileSync(
    path.join(DATA_DIR, "redirects.json"),
    JSON.stringify(redirects, null, 2),
    "utf-8"
  );

  console.log("\nSlug mapping applied:");
  console.log(`  Renamed: ${renamed}`);
  console.log(`  Unchanged: ${skipped}`);
  console.log(`  Not found: ${notFound.length}`);
  console.log(`  Redirect entries: ${Object.keys(redirects).length}`);

  if (notFound.length > 0) {
    console.log("\nCould not find files for:");
    for (const url of notFound.slice(0, 10)) {
      console.log(`  - ${url}`);
    }
    process.exit(1);
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
